kamu-adapter-http/src/http_server_simple_transfer_protocol.rs

```diff
@@ -0,0 +1,229 @@
+            return Err(axum::http::StatusCode::INTERNAL_SERVER_ERROR);
+        }
+    };
+
+    let body = axum_extra::body::AsyncReadBody::new(checkpoint_stream);
+    Ok(axum::response::Response::builder()
+        .body(axum::body::boxed(body))
+        .unwrap())
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+pub async fn dataset_push_ws_upgrade_handler(
+    ws: axum::extract::ws::WebSocketUpgrade,
+    dataset: Extension<Arc<dyn Dataset>>,
+) -> axum::response::Response {
+    ws.on_upgrade(|socket| {
+        ws_smart_transfer_protocol_axum_server::dataset_push_ws_handler(socket, dataset.0)
+    })
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+pub async fn dataset_pull_ws_upgrade_handler(
+    ws: axum::extract::ws::WebSocketUpgrade,
+    dataset: Extension<Arc<dyn Dataset>>,
+    axum::TypedHeader(api_host): axum::TypedHeader<axum::headers::Host>,
+    axum::extract::Path(params): axum::extract::Path<HashMap<String, String>>,
+) -> axum::response::Response {
+    let base_url = resolve_base_api_url(api_host);
+
+    let dataset_name_param = params.get(PARAMETER_DATASET_NAME).unwrap();
+    let dataset_url = base_url
+        .join((String::from(dataset_name_param) + "/").as_str())
+        .unwrap();
+
+    ws.on_upgrade(move |socket| {
+        ws_smart_transfer_protocol_axum_server::dataset_pull_ws_handler(
+            socket,
+            dataset.0,
+            dataset_url,
+        )
+    })
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+pub async fn resolve_dataset_by_name(
+    dataset_resolver: axum::Extension<Arc<dyn DatasetResolver>>,
+    axum::extract::Path(params): axum::extract::Path<HashMap<String, String>>,
+    mut req: axum::http::Request<axum::body::Body>,
+    next: axum::middleware::Next<axum::body::Body>,
+) -> Result<axum::response::Response, axum::http::StatusCode> {
+    let dataset_name_param = params.get(PARAMETER_DATASET_NAME).unwrap();
+    let dataset_resolution = dataset_resolver
+        .resolve_dataset(dataset_name_param.as_str())
+        .await;
+
+    match dataset_resolution {
+        Ok(dataset) => {
+            req.extensions_mut().insert(dataset);
+            Ok(next.run(req).await)
+        }
+        Err(e) => match e {
+            kamu::domain::GetDatasetError::NotFound(_) => Err(axum::http::StatusCode::NOT_FOUND),
+            kamu::domain::GetDatasetError::Internal(_) => {
+                tracing::debug!(
+                    "Internal error while resolving dataset '{}'",
+                    dataset_name_param
+                );
+                Err(axum::http::StatusCode::INTERNAL_SERVER_ERROR)
+            }
+        },
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+fn resolve_base_api_url(api_host: axum::headers::Host) -> Url {
+    let mut base_url_str = String::from("http://");
```

Comment:
Note that hard-coded HTTP will fail once we deploy on AWS.

Reply:
The easiest fix would be to configure `base_url` from the outside.

Alternatively we could try writing an extractor similar to `axum::extract::Host` to guess the scheme e.g. from `X-Forwarded-Proto` and local port (80 / 443).

Reply:
![image](https://user-images.githubusercontent.com/8875909/228237055-e98d2454-e4d6-4a62-8333-2fffc455d7f8.png)

Reply:
Added `KAMU_PROTOCOL_SCHEME` env var